src/providers/MeetingProvider/MeetingManager.ts

```diff
@@ -269,31 +276,39 @@ export class MeetingManager implements AudioVideoObserver {
       (await this.audioVideo?.listAudioOutputDevices()) || [];
   }
 
-  setupDeviceLabelTrigger(): void {
-    const callback = async (): Promise<MediaStream> => {
-      this.devicePermissionStatus = DevicePermissionStatus.IN_PROGRESS;
-      this.publishDevicePermissionStatus();
-      try {
-        const devices = await navigator.mediaDevices.enumerateDevices();
-        const hasVideoInput = devices.some(value => {
-          return value.kind === 'videoinput';
-        });
+  setupDeviceLabelTrigger(deviceLabels: DeviceLabels | DeviceLabelTrigger = DeviceLabels.AudioAndVideo): void {
+    let callback: () => Promise<MediaStream>;
 
-        const stream = await navigator.mediaDevices.getUserMedia({
-          audio: true,
-          video: hasVideoInput
-        });
-
-        this.devicePermissionStatus = DevicePermissionStatus.GRANTED;
-        this.publishDevicePermissionStatus();
-        return stream;
-      } catch (error) {
-        console.error('Failed to get device permissions');
-        this.devicePermissionStatus = DevicePermissionStatus.DENIED;
+    if (typeof deviceLabels === "function") {
```

Comment:
TypeScript should make the following work:

```typescript
switch deviceLabels {
  case DeviceLabels.None: {
    …
    break;
  }
…

  case DeviceLabels.Audio: {
  }

…

  default: {
    // I must be a function.
  }
}
```

See [this Playground](https://www.typescriptlang.org/play?#code/KYOwrgtgBAYg9nKBvAUFKAjAhgJwDRqZYBeKAvigGZggDGALgJZwhQAeAFAG4BcsCUAD5QOHAJRQAvAD4oXOIwAmYsX3lLkhAM4B3RvVoALEVwmp06Wli3B+cAHTYcfDDmBYA1gG5Cl67fgHbGIXN08fCyhFYEosMAAbeh5fCy5xCKgKMiA).

Reply:
I'm not sure I follow, it seems the `switch case` is not good for our cases. What we need to do is:
- `DeviceLabels` type
  1. Config the `constraints` of `getUserMedia()` based on `DeviceLabels`
  2. Set the callback with constraints
- `DeviceLabelTrigger` type
  1. Set the user passed callback

So the code is more like:
``` ts
let callback: () => Promise<MediaStream>;

switch (deviceLabels) {
  case DeviceLabels.None: {
    callback = () => Promise.resolve(new MediaStream());
    break;
  }
  case DeviceLabels.Audio: {
  // config constraints of `getUserMedia`
  }
  case DeviceLabels.Video: {
  // config constraints of `getUserMedia`
  }
  case DeviceLabels.AudioAndVideo: {
  // config constraints of `getUserMedia`
  }

  // If the `deviceLables` is `DeviceLabels` type, we need to assign 
  // the device label trigger with `constraints` to `callback`, and break here

  default: {
    // I must be a function.
    callback = deviceLabels;
  }

  this.audioVideo?.setDeviceLabelTrigger(callback);
}
```
Or it could be quite redundant, if we set the callback in each case and break.

Reply:
Redundancy is OK! Optimize for clarity and code changing over time.

Imagine if another engineer needs to add a case to `DeviceLabels` in the future. Using a comprehensive `switch` statement will guarantee you a compiler error if you fail to handle the new case, which is great. `if` doesn't get you that all the time. You can do a `"function"` type check then a `switch` afterwards, if you prefer. Use your judgment — as requested, I was just suggesting alternative possibilities.

Reply:
Thank you for such a detailed explanation, I appreciate it. I will keep it in mind!